xblock/fields.py

```diff
@@ -838,31 +842,27 @@ def from_json(self, value):
         """
         Parse the date from an ISO-formatted date string, or None.
         """
-        if isinstance(value, basestring):
+        if value is None:
+            return None
 
+        if isinstance(value, basestring):
             # Parser interprets empty string as now by default
             if value == "":
                 return None
-
             try:
-                parsed_date = dateutil.parser.parse(value)
+                value = dateutil.parser.parse(value)
             except (TypeError, ValueError):
                 raise ValueError("Could not parse {} as a date".format(value))
 
-            if parsed_date.tzinfo is not None:  # pylint: disable=maybe-no-member
-                parsed_date.astimezone(pytz.utc)  # pylint: disable=maybe-no-member
-            else:
-                parsed_date = parsed_date.replace(tzinfo=pytz.utc)  # pylint: disable=maybe-no-member
-
-            return parsed_date
-
-        if value is None:
-            return None
-
-        if isinstance(value, datetime.datetime):
-            return value
+        if not isinstance(value, datetime.datetime):
+            raise TypeError(
+                "Value should be loaded from a string, a datetime object or None, not {}".format(type(value))
+            )
 
-        raise TypeError("Value should be loaded from a string, not {}".format(type(value)))
+        if value.tzinfo is not None:  # pylint: disable=maybe-no-member
+            return value.astimezone(pytz.utc)  # pylint: disable=maybe-no-member
```

Comment:
The main problem with the old implementation was that it correctly adjusted the timezone of values passed in as string, but did not adjust the timezone of values passed in as `datetime` instances.  This results in rather inconsistent behaviour, since `yaml.safe_load()`, which is used as first step when deserialising data, already parses unquoted date strings into `datetime` instances, whereas quoted date string are returned as strings, which will only be parsed into `datetime` instances in this function.  In essence, adding and removing quotes resulted in different handling of timezones.  I don't think there is any code out there relying on this bug, but I've been wrong with this kind of assumption before (see https://xkcd.com/1172/).  For what it's worth, all edx-platform tests pass with this patch applied.

@doctoryes I think `.astimezone()` did change the timezone in some cases even before this change, but it wasn't applied when the value passed in already was a `datetime` instance.

@sarina Not sure about this.  This fix doesn't change the treatment of timezones fundamentally; it only makes it behave consistently.  Maybe edx-platform only works correctly if the server time is UTC?  I really don't know.